### PR TITLE
feat: do not serialize functions or constructors that are not invoked.

### PR DESCRIPTION
**Pull request by thantos**
```ts
class C {
   static S = "x"
   constructor() {}
}
```

```ts
() => {
   // does not need the whole class
   C.S
}
```

```ts
() => {
   // needs the whole class
   new C()
}
```

```ts
() => {
   // needs the whole class
   C.S
   new C();
}
```

```ts
() => {
   // needs the whole class
   const doSomething = (d: any) => {}
   doSomething(C);
}
```

### CODE REVIEW

File: src/closure/createClosure.ts [hunk 1]
```diff
@@ -635,10 +642,11 @@ async function analyzeFunctionInfoAsync(
         logInfo
       );
 
-      // try to only serialize out the properties that were used by the user's code.
+      // try to only serialize out the pro  perties that were used by the user's code.
```
Comment (sam-goodwin) [inline src/closure/createClosure.ts]: nit: typo

File: src/closure/createClosure.ts [hunk 2]
```diff
@@ -1038,18 +1088,34 @@ async function getOrCreateEntryAsync(
...
+      // if not invoked, let the property be treated like an object
+      let serializeAll = true;
+
+      if (
+        typeof isInvoked === "boolean" &&
+        !isInvoked &&
+        capturedObjectProperties &&
+        capturedObjectProperties.length > 0
+      ) {
+        entry.object = entry.object || { env: new Map() };
```
Comment (sam-goodwin) [inline src/closure/createClosure.ts]: Do you mean to use `??` instead of `||`?

Comment (sam-goodwin) [reply src/closure/createClosure.ts]: Why `new Map()`?

File: src/closure/createClosure.ts [hunk 3]
```diff
@@ -1038,18 +1088,34 @@ async function getOrCreateEntryAsync(
...
       await captureModuleAsync(normalizedModuleName);
     } else if (obj instanceof Function) {
-      // Serialize functions recursively, and store them in a closure property.
-      entry.function = await analyzeFunctionInfoAsync(
-        obj,
-        context,
-        serialize,
-        logInfo
-      );
+      // if not invoked, let the property be treated like an object
```
Comment (sam-goodwin) [inline src/closure/createClosure.ts]: meaning attempt to only serialize used properties?

Comment (sam-goodwin) [reply src/closure/createClosure.ts]: i.e. tree-shake?

**Review by sam-goodwin — APPROVED** [review]
Great job.

File: src/closure/createClosure.ts (hunk 2 above)
Comment (thantos) [reply src/closure/createClosure.ts]: I copies this from elsewhere in the code, lol, but it appears that object needs to be a map or else things start to fail.

File: src/closure/createClosure.ts (hunk 3 above)
Comment (thantos) [reply src/closure/createClosure.ts]: right, previous behavior was that anything recognized as a function `instanceof Function` would be fully serialized, meaning static properties made a mess.

Comment (thantos) [reply src/closure/createClosure.ts]: And namespaces more so :-)

**Pull request merged**